File: frontend/src/api/annotation_list.ts
import { apiClient } from "./utils";

// Types matching the backend schemas
export interface AnnotationListUploadResponse {
  success: boolean;
  message: string;
  total_records_created: number;
  record_ids: string[];
  root_type: string;
}

export interface CategoryOutput {
  id?: string;
  name: string;
  description?: string;
  level?: number;
  parent?: string;
  mnemonic?: string;
  examples?: string[];
  notes?: string;
  subcategories?: CategoryOutput[];
}

export interface HierarchicalJSONOutput {
  version?: string;
  title: string;
  description?: string;
  copyright?: string;
  categories: CategoryOutput[];
}

export interface AnnotationListResponse {
  id: string;
  type?: string;
  type_id?: string;
  title: string;
  level?: string;
  parent_id?: string;
  description?: string;
  created_by?: string;
  meta?: Record<string, any>;
  created_at: string;
  updated_at?: string;
}

export interface AnnotationListCreate {
  title: string;
  type?: string;
  type_id?: string;
  level?: string;
  parent_id?: string;
  description?: string;
  meta?: Record<string, any>;
}

export interface AnnotationListUpdate {
  title?: string;
  level?: string;
  parent_id?: string;
  description?: string;
  meta?: Record<string, any>;
}

export interface AnnotationListFilters {
  title?: string;
  type?: string;
  created_by?: string;
  created_at?: string;
  skip?: number;
  limit?: number;
  [key: string]: string | number | boolean | undefined;
}

export const annotationListApi = {
  /**
   * Upload a JSON file with hierarchical annotation list
   */
  uploadFile: async (file: File): Promise<AnnotationListUploadResponse> => {
    const formData = new FormData();
    formData.append("file", file);
    
    return await apiClient.post<AnnotationListUploadResponse>(
      "/annotation-lists/upload",
      formData
    );
  },

  /**
   * Get annotation lists by type in hierarchical format
   */
  getByTypeHierarchical: async (type_id: string): Promise<HierarchicalJSONOutput> => {
    return await apiClient.get<HierarchicalJSONOutput>(
      `/annotation-lists/type/${encodeURIComponent(type_id)}`
    );
  },

  getAll: async (params?: AnnotationListFilters): Promise<AnnotationListResponse[]> => {
    return await apiClient.get<AnnotationListResponse[]>(
      `/annotation-lists/`,
      params
    );
  },

  /**
   * Delete all annotation lists by type
   */
  deleteByType: async (type_id: string): Promise<{ success: boolean; message: string; deleted_count: number }> => {
    return await apiClient.delete(
      `/annotation-lists/type/${encodeURIComponent(type_id)}`
    );
  },

  /**
   * Helper function to validate file type
   */
  validateFileType: (file: File): boolean => {
    return file.name.endsWith(".json");
  },

  /**
   * Helper function to read and validate JSON structure
   */
  validateJsonStructure: async (file: File): Promise<{ valid: boolean; error?: string }> => {
    try {
      const content = await file.text();
      const parsed = JSON.parse(content);

      // Check for required fields
      if (!parsed.title || typeof parsed.title !== "string") {
        return { valid: false, error: 'Missing or invalid "title" field' };
      }

      if (!parsed.categories || !Array.isArray(parsed.categories)) {
        return { valid: false, error: 'Missing or invalid "categories" array' };
      }

      if (parsed.categories.length === 0) {
        return { valid: false, error: "Categories array is empty" };
      }

      // Validate first category has required fields
      const firstCategory = parsed.categories[0];
      if (!firstCategory.name) {
        return { valid: false, error: 'Categories must have a "name" field' };
      }

      return { valid: true };
    } catch (error) {
      return {
        valid: false,
        error: error instanceof Error ? error.message : "Invalid JSON format",
      };
    }
  },

  /**
   * Helper function to format file size
   */
  formatFileSize: (bytes: number): string => {
    if (bytes === 0) return "0 Bytes";

    const k = 1024;
    const sizes = ["Bytes", "KB", "MB", "GB"];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + " " + sizes[i];
  },

  /**
   * Helper function to count total items in hierarchy
   */
  countHierarchyItems: (categories: CategoryOutput[]): number => {
    let count = categories.length;
    categories.forEach(cat => {
      if (cat.subcategories) {
        count += annotationListApi.countHierarchyItems(cat.subcategories);
      }
    });
    return count;
  },

  /**
   * Create a new annotation list item
   */
  createItem: async (item: AnnotationListCreate): Promise<AnnotationListResponse> => {
    return await apiClient.post<AnnotationListResponse>("/annotation-lists/", item);
  },

  /**
   * Update an annotation list item
   */
  updateItem: async (itemId: string, item: AnnotationListUpdate): Promise<AnnotationListResponse> => {
    return await apiClient.put<AnnotationListResponse>(`/annotation-lists/${encodeURIComponent(itemId)}`, item);
  },

  /**
   * Delete an annotation list item
   */
  deleteItem: async (itemId: string): Promise<{ success: boolean; message: string }> => {
    return await apiClient.delete(`/annotation-lists/${encodeURIComponent(itemId)}`);
  },

  /**
   * Get an annotation list item by ID
   */
  getItem: async (itemId: string): Promise<AnnotationListResponse> => {
    return await apiClient.get<AnnotationListResponse>(`/annotation-lists/${encodeURIComponent(itemId)}`);
  },
};

File: backend/schemas/annotation_list.py
from pydantic import BaseModel, ConfigDict, Field
from typing import Optional, Dict, Any, List
from datetime import datetime


class AnnotationListBase(BaseModel):
    """Base schema for annotation list."""
    title: str
    type: Optional[str] = None
    level: Optional[str] = None
    description: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None


class AnnotationListCreate(AnnotationListBase):
    """Schema for creating an annotation list item."""
    parent_id: Optional[str] = None
    type_id: Optional[str] = None


class AnnotationListUpdate(BaseModel):
    """Schema for updating an annotation list item."""
    title: Optional[str] = None
    level: Optional[str] = None
    description: Optional[str] = None
    parent_id: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None


class AnnotationListResponse(AnnotationListBase):
    """Schema for annotation list response."""
    model_config = ConfigDict(from_attributes=True)
    
    id: str
    type_id: Optional[str] = None
    parent_id: Optional[str] = None
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    @classmethod
    def model_validate(cls, obj, **kwargs):
        """Custom validation to populate type from annotation_type relationship."""
        if hasattr(obj, 'annotation_type') and obj.annotation_type:
            # Create a dict with all attributes plus the type from relationship
            data = {
                'id': obj.id,
                'type_id': obj.type_id,
                'title': obj.title,
                'type': obj.annotation_type.name,
                'level': obj.level,
                'description': obj.description,
                'meta': obj.meta,
                'parent_id': obj.parent_id,
                'created_by': obj.created_by,
                'created_at': obj.created_at,
                'updated_at': obj.updated_at
            }
            return super().model_validate(data, **kwargs)
        return super().model_validate(obj, **kwargs)


class CategoryInput(BaseModel):
    """Schema for category in hierarchical JSON input."""
    id: Optional[str] = None
    name: str
    description: Optional[str] = None
    level: Optional[int] = None
    parent: Optional[str] = None
    mnemonic: Optional[str] = None
    examples: Optional[List[Any]] = None
    notes: Optional[str] = None
    subcategories: Optional[List['CategoryInput']] = None
    
    model_config = ConfigDict(extra='allow')  # Allow extra fields


class HierarchicalJSONInput(BaseModel):
    """Schema for hierarchical JSON input."""
    version: Optional[str] = None
    title: str
    description: Optional[str] = None
    copyright: Optional[str] = None
    categories: List[CategoryInput]
    
    model_config = ConfigDict(extra='allow')  # Allow extra fields


class AnnotationListBulkCreateRequest(BaseModel):
    """Schema for bulk create request."""
    data: HierarchicalJSONInput = Field(..., description="Hierarchical JSON data")


class AnnotationListBulkCreateResponse(BaseModel):
    """Schema for bulk create response."""
    success: bool
    message: str
    total_records_created: int
    record_ids: List[str]
    root_type: str


class CategoryOutput(BaseModel):
    """Schema for category in hierarchical JSON output."""
    id: Optional[str] = None
    name: str
    description: Optional[str] = None
    level: Optional[int] = None
    parent: Optional[str] = None
    mnemonic: Optional[str] = None
    examples: Optional[List[Any]] = None
    notes: Optional[str] = None
    subcategories: Optional[List['CategoryOutput']] = None
    
    model_config = ConfigDict(extra='allow')  # Allow extra fields


class HierarchicalJSONOutput(BaseModel):
    """Schema for hierarchical JSON output."""
    version: Optional[str] = None
    title: str
    description: Optional[str] = None
    copyright: Optional[str] = None
    categories: List[CategoryOutput]
    
    model_config = ConfigDict(extra='allow')  # Allow extra fields


# Update forward references for recursive models
CategoryInput.model_rebuild()
CategoryOutput.model_rebuild()

